Add saveOnArrival support

eschramm opened the pull request:
Supports https://github.com/CareEvolution/MyDataHelps-iOS/issues/1086. At the ResearchKit level, this PR serves to simply:
- add a BOOL for `saveOnArrival` to the ORK[1]Step
- remove the back button to prevent backwards navigation on a step with `saveOnArrival == YES`
- remove the cancel button on a step with `saveOnArrival == YES`
- prevent navigation beyond a step with `saveOnArrival == YES` - essentially this must be the last step of a survey

### Testing
No testing necessarily needed at this level, see https://github.com/CareEvolution/MyDataHelps-iOS/pull/1152.

eschramm committed:
add saveOnArrival support

eschramm committed:
hide cancel button if saveOnArrival

eschramm committed:
prevent navigation beyond saveOnArrival

eschramm committed:
improved handling for skipRules

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1Step.m:
```diff
@@ -114,12 +115,13 @@ - (BOOL)isEqual:(id)object {
 - (NSUInteger)hash {
     // Ignore the task reference - it's not part of the content of the step.
-    return _identifier.hash ^ _title.hash ^ _text.hash ^ (_optional ? 0xf : 0x0) ^ _nextButtonText.hash ^ _userInfo.hash;
+    return _identifier.hash ^ _title.hash ^ _text.hash ^ (_optional ? 0xf : 0x0) ^ _nextButtonText.hash ^ _userInfo.hash ^ (_saveOnArrival ? 0xf : 0x1);
```
Minor: based on examples such as [ORK1ConsentReviewStep](https://github.com/CareEvolution/ResearchKit/blob/43c3f48dc4fe82f6e352bcbdb96395e2efde292f/ORK1Kit/ORK1Kit/Consent/ORK1ConsentReviewStep.m#L119), looks like the pattern here for hashing multiple bools would be for saveOnArrival to use `0xe : 0x1`

eschramm replied:
Good catch - I think I understand this better now - each case should be a separate, unique "bit flag"

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m:
```diff
@@ -1202,7 +1202,7 @@ - (ORK1StepViewController *)viewControllerForStep:(ORK1Step *)step {
     
     
-    if (stepViewController.cancelButtonItem == nil) {
+    if (stepViewController.cancelButtonItem == nil && !stepViewController.step.saveOnArrival) {
```
`ORK1TaskViewController.applicationFinishedRestoringState` has the same logic of `if (stepViewController.cancelButtonItem == nil) { stepViewController.cancelButtonItem = [self defaultCancelButtonItem]; }`. Any reason not to add the `!stepViewController.step.saveOnArrival` there also?

eschramm replied:
Good catch

mmertsock commented on ResearchKit/Common/ORKStep.m:
```diff
@@ -114,12 +115,13 @@ - (BOOL)isEqual:(id)object {
 - (NSUInteger)hash {
     // Ignore the task reference - it's not part of the content of the step.
-    return _identifier.hash ^ _title.hash ^ _text.hash ^ (_optional ? 0xf : 0x0) ^ _nextButtonText.hash ^ _userInfo.hash;
+    return _identifier.hash ^ _title.hash ^ _text.hash ^ (_optional ? 0xf : 0x0) ^ _nextButtonText.hash ^ _userInfo.hash ^ (_saveOnArrival ? 0xf : 0x1);;
```
Same comment here as in ORK1Step.m

mmertsock commented on ResearchKit/Common/ORKTaskViewController.m:
```diff
@@ -1287,7 +1287,7 @@ - (ORKStepViewController *)viewControllerForStep:(ORKStep *)step {
     
     
-    if (stepViewController.cancelButtonItem == nil) {
+    if (stepViewController.cancelButtonItem == nil && !stepViewController.step.saveOnArrival) {
```
Same comment here as in ORK1TaskViewController.m

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1NavigableOrderedTask.m:
```diff
@@ -174,7 +180,7 @@ - (ORK1Step *)stepAfterStep:(ORK1Step *)step withResult:(ORK1TaskResult *)result
         ORK1SkipStepNavigationRule *skipNavigationRule = _skipStepNavigationRules[nextStep.identifier];
         if ([skipNavigationRule stepShouldSkipWithTaskResult:result]) {
-            return [self stepAfterStep:nextStep withResult:result];
+            return [self stepAfterStep:nextStep withResult:result ignoreSaveOnArrival:YES];
```
Checking my understanding:

- Given steps A, B, and C, where B.saveOnArrival = YES
- There's a skip navigation rule that indicates step B should be skipped
- `stepAfterStep:A` is called

At this point in the code, `nextStep` has been set to `B`, but we need to find the step after B since it's being skipped. Thus we call `[stepAfterStep:B ignoreSaveOnArrival:YES]`, which determines that C should be the next step?

eschramm replied:
That is correct. In early testing of the prototype, I noticed it would auto-complete (advance and return nil) if there were a skip rule on the last step. This is a way to pass on a little state recursively to handle that case.

mmertsock reviewed (commented):
Overall design LGTM, just some minor questions/comments

eschramm committed:
PR fixes

- adjusted hashing bit flags to be unique
- additional case to check for saveOnArrival when looking to add cancel button

mmertsock reviewed (approved):
LGTM 🚢

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1Step.h:
```diff
@@ -247,6 +247,15 @@ The default value of this property is 'nil'.
 
+/**
+ When YES, upon loading this step, the ORK1TaskViewController will instruct
+ the delegate to save the task results.
```
Forgot to mention some notes about this documentation:

- ResearchKit itself doesn't have any intrinsic concept of "saving task results" in the way that you mean here; and existing mentions of "save" elsewhere in ORK1TaskViewController's documentation tends to refer to updating the ORK1TaskResult with each step, or persisting restorationData, rather than handling the final survey result after completion.
- There is no new delegate method specific to saveOnArrival, so the "will instruct the delegate" phrasing is inaccurate.

The documentation for the `didFinishWithReason` delegate method states:

>  In most circumstances, the receiver should dismiss the task view controller in response to this method, and may also need to collect and process the results of the task.

So perhaps the `saveOnArrival` documentation should read more like that, and emphasize that the app is intended to instead perform that "collect and process" behavior upon reaching the step, e.g. in `stepViewControllerWillAppear`, and only perform dismissal or error handling behaviors in `didFinishWithReason`.

And could also mention that if the user arrives at any step with `saveOnArrival`, it will be the final step presented to the user and the task's `result` will be complete.

eschramm replied:
Good point - have made https://github.com/CareEvolution/ResearchKit/pull/131 to address.
